src/main/java/org/opensearch/knn/training/ByteTrainingDataConsumer.java

```diff
@@ -0,0 +1,90 @@
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.training;
+
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.knn.jni.JNICommons;
+import org.opensearch.knn.index.memory.NativeMemoryAllocation;
+import org.opensearch.search.SearchHit;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Transfers byte vectors from JVM to native memory.
+ */
+public class ByteTrainingDataConsumer extends TrainingDataConsumer<byte[]> {
+
+    /**
+     * Constructor
+     *
+     * @param trainingDataAllocation NativeMemoryAllocation that contains information about native memory allocation.
+     */
+    public ByteTrainingDataConsumer(NativeMemoryAllocation.TrainingDataAllocation trainingDataAllocation) {
+        super(trainingDataAllocation);
+    }
+
+    @Override
+    public void accept(List<byte[]> byteVectors) {
+        long memoryAddress = trainingDataAllocation.getMemoryAddress();
+        memoryAddress = JNICommons.storeByteVectorData(memoryAddress, byteVectors.toArray(new byte[0][0]), byteVectors.size());
+        trainingDataAllocation.setMemoryAddress(memoryAddress);
+    }
+
+    @Override
+    public List<byte[]> extractVectorsFromHits(SearchResponse searchResponse, int vectorsToAdd, String fieldName) {
+        SearchHit[] hits = searchResponse.getHits().getHits();
+        List<byte[]> byteVectors = new ArrayList<>();
+        String[] fieldPath = fieldName.split("\\.");
+        int nullVectorCount = 0;
+
+        for (int vector = 0; vector < vectorsToAdd; vector++) {
+            Object fieldValue = extractFieldValue(hits[vector], fieldPath);
+            if (fieldValue == null) {
+                nullVectorCount++;
+                continue;
+            }
+
+            byte[] byteArray;
+            if (!(fieldValue instanceof List<?>)) {
+                continue;
+            }
+            List<Number> fieldList = (List<Number>) fieldValue;
+            byteArray = new byte[fieldList.size()];
+            for (int i = 0; i < fieldList.size(); i++) {
+                byteArray[i] = fieldList.get(i).byteValue();
+            }
+
+            byteVectors.add(byteArray);
+        }
+
+        if (nullVectorCount > 0) {
+            logger.warn("Found {} documents with null byte vectors in field {}", nullVectorCount, fieldName);
+        }
+
+        return byteVectors;
+    }
+
+    @Override
+    public void processTrainingVectors(SearchResponse searchResponse, int vectorsToAdd, String fieldName) {
+        List<byte[]> vectors = extractVectorsFromHits(searchResponse, vectorsToAdd, fieldName);
+
+        setTotalVectorsCountAdded(getTotalVectorsCountAdded() + vectors.size());
+
+        accept(vectors);
+    }
+
+    @Override
+    public List<byte[]> getTotalAddedVectors() {
```

Comment:
I see this method is called below. Where the implementation is?

Reply:
this method is used in TestFloatTrainingDataConsumer
```
    // create test float training data consumer class extending FloatTrainingDataConsumer
    private static class TestFloatTrainingDataConsumer extends FloatTrainingDataConsumer {
        @Getter
        private List<Float[]> totalAddedVectors = new ArrayList<>();

        public TestFloatTrainingDataConsumer(NativeMemoryAllocation.TrainingDataAllocation trainingDataAllocation) {
            super(trainingDataAllocation);
        }

        @Override
        public void processTrainingVectors(SearchResponse searchResponse, int vectorsToAdd, String fieldName) {
            List<Float[]> vectors = extractVectorsFromHits(searchResponse, vectorsToAdd, fieldName);
            totalAddedVectors.addAll(vectors);
            setTotalVectorsCountAdded(getTotalVectorsCountAdded() + vectors.size());
            accept(vectors);
        }
    }
```

Reply:
It is wired to have a method only to be used in testing... Can we do differently?

Reply:
I tried but couldn't find other way to read the whole vectors

Reply:
You don't need to. For unit test, all you need to verify is that JNILayer method is called properly with expected parameter.

Reply:
updated